solvers/H3LAPD/EquationSystems/DriftReducedSystem.cpp

```diff
@@ -0,0 +1,578 @@
+   * temporary output array
+   */
+  Array<OneD, MR::ExpListSharedPtr> tmp_fields(nfields);
+  Array<OneD, Array<OneD, NekDouble>> tmp_inarray(nfields);
+  Array<OneD, Array<OneD, NekDouble>> tmp_outarray(nfields);
+  for (auto ii = 0; ii < nfields; ii++) {
+    int idx = m_field_to_index.get_idx(field_names[ii]);
+    tmp_fields[ii] = m_fields[idx];
+    tmp_inarray[ii] = Array<OneD, NekDouble>(npts);
+    Vmath::Vcopy(npts, in_arr[idx], 1, tmp_inarray[ii], 1);
+    tmp_outarray[ii] = Array<OneD, NekDouble>(out_arr[idx].size());
+  }
+  // Compute advection terms; result is returned in temporary output array
+  adv_obj->Advect(tmp_fields.size(), tmp_fields, adv_vel, tmp_inarray,
+                  tmp_outarray, time);
+
+  // Subtract temporary output array from the appropriate indices of out_arr
+  for (auto ii = 0; ii < nfields; ii++) {
+    int idx = m_field_to_index.get_idx(eqn_labels[ii]);
+    Vmath::Vsub(out_arr[idx].size(), out_arr[idx], 1, tmp_outarray[ii], 1,
+                out_arr[idx], 1);
+  }
+}
+
+/**
+ * @brief Add (density) source term via a Nektar session function.
+ *
+ * @details Looks for a function called "dens_src", evaluates it, and adds the
+ * result to @p out_arr
+ *
+ * @param[out] out_arr RHS array to add the source too
+ * @todo Check function exists, rather than relying on Nektar ASSERT
+ */
+void DriftReducedSystem::add_density_source(
+    Array<OneD, Array<OneD, NekDouble>> &out_arr) {
+
+  int ne_idx = m_field_to_index.get_idx("ne");
+  int npts = GetNpoints();
+  Array<OneD, NekDouble> tmpx(npts), tmpy(npts), tmpz(npts);
+  m_fields[ne_idx]->GetCoords(tmpx, tmpy, tmpz);
+  Array<OneD, NekDouble> dens_src(npts, 0.0);
+  LU::EquationSharedPtr dens_src_func =
+      m_session->GetFunction("dens_src", ne_idx);
+  dens_src_func->Evaluate(tmpx, tmpy, tmpz, dens_src);
+  Vmath::Vadd(npts, out_arr[ne_idx], 1, dens_src, 1, out_arr[ne_idx], 1);
+}
+
+/**
+ * @brief Adds particle sources.
+ * @details For each <field_name> in @p target_fields , look for another field
+ * called <field_name>_src. If it exists, add the physical values of
+ * field_name_src to the appropriate index of @p out_arr.
+ *
+ * @param target_fields list of Nektar field names for which to look for a
+ * '_src' counterpart
+ *  @param[out] out_arr      the RHS array
+ *
+ */
+void DriftReducedSystem::add_particle_sources(
+    std::vector<std::string> target_fields,
+    Array<OneD, Array<OneD, NekDouble>> &out_arr) {
+  for (auto target_field : target_fields) {
+    int src_field_idx = m_field_to_index.get_idx(target_field + "_src");
+
+    if (src_field_idx >= 0) {
+      auto tmp_it = std::find(m_int_fld_names.cbegin(), m_int_fld_names.cend(),
+                              target_field);
+      ASSERTL0(tmp_it != m_int_fld_names.cend(),
+               "Target field for particle source ['" + target_field +
+                   "'] term not recognised.")
+      auto field_idx = std::distance(m_int_fld_names.cbegin(), tmp_it);
```

Comment:
The source fields arn't indexed by the field_to_index class?

Reply:
No, but I agree it's far from obvious what's going on there - I've added some comments to try and clarify.